### DISCUSSION TITLE
How to get empty paragraphs when using `mdast-util-from-markdown`

### DISCUSSION THREAD

**Discussion by nicknikolov** [syntax-tree/unist#33] [Q&A] [answered]
I use mdast-util-from-markdown through remark-parse and I'd like to get the empty paragraph in my AST.
E.g. when using parsing something like
```

foo

baa

```

I get 
```
{
  "type": "root",
  "children": [
    {
      "type": "paragraph",
      "children": [
        {
          "type": "text",
          "value": "foo",
          "position": {
            "start": {
              "line": 2,
              "column": 1,
              "offset": 1
            },
            "end": {
              "line": 2,
              "column": 4,
              "offset": 4
            }
          }
        }
      ],
      "position": {
        "start": {
          "line": 2,
          "column": 1,
          "offset": 1
        },
        "end": {
          "line": 2,
          "column": 4,
          "offset": 4
        }
      }
    },
    {
      "type": "paragraph",
      "children": [
        {
          "type": "text",
          "value": "bar",
          "position": {
            "start": {
              "line": 4,
              "column": 1,
              "offset": 6
            },
            "end": {
              "line": 4,
              "column": 4,
              "offset": 9
            }
          }
        }
      ],
      "position": {
        "start": {
          "line": 4,
          "column": 1,
          "offset": 6
        },
        "end": {
          "line": 4,
          "column": 4,
          "offset": 9
        }
      }
    }
  ],
  "position": {
    "start": {
      "line": 1,
      "column": 1,
      "offset": 0
    },
    "end": {
      "line": 6,
      "column": 1,
      "offset": 11
    }
  }
}
```

So the empty paragraphs around my text get "lost" but I would like to use them to visualise the markdown text as is (including the empty paragraphs on line 1 3 and 5). 

What's the best way achieve this?

PS: Hopefully this is the most appropriate place to ask, if not let me know! Thanks

**Comment by wooorm** [comment, marked as answer]
There are no empty paragraphs in markdown 🤷‍♂️

Could you expand more on *what* and *why* you want this?

**Reply by nicknikolov** [reply]
I see! Bummer. I just want to render the markdown as is on my client. Currently all these empty lines in my markdown files disappear. Perhaps I can write some "post-processing" code to find lines that don't have an AST node and add one as empty... any other ideas?

Thanks for the quick reply!

**Reply by ChristianMurphy** [reply]
> I just want to render the markdown as is on my client.

remark doesn't remove the newline, take for example this roundtrip of markdown through remark
https://astexplorer.net/#/gist/f56e1cbf69bb681bdbc4162d36742472/a3533dd8d55e69c9fecb5476e6bfeb52ab4e3921

> any other ideas?

It sounds like you may be changing from markdown to another format? (HTML?)
Or have applied some plugins?

**Reply by nicknikolov** [reply]
Hm. Yes I use https://github.com/inokawa/remark-slate-transformer 

So from your example, my problem is line 2 and 4 don't have a paragraph associated so I basically "lose" those lines when I transform. But since the information is kept as line numbers they are still there, I just have to treat them as a special case.

**Reply by ChristianMurphy** [reply]
You don't lose them, having an mdast paragraph implies a line after.
The mdast to slate transform is choosing whether or not to include that newline on the slate side, not mdast.

**Reply by wooorm** [reply]
@nicknikolov you seem to be missing what markdown is. It's not plain text. It's an alternative to html. That blank line is not an empty paragraph, it's the split between two paragraphs.

Again: could you expand on what you want to do, and why you want to do it like that?

**Reply by nicknikolov** [reply]
I see. I think you're spot on and I have a conceptual misunderstanding.

As to my use case, it's pretty simple. I have a markdown editor on the client. I just want to load markdown files and preserve the structure, so
```
foo 

bar
```
stays as is, rather than becoming
```
foo
bar
```

But I suppose, judging by what you're telling me, this is a problem with the transformer rather than the markdown AST?

**Reply by nicknikolov** [reply]
Oops I missed Christian's comment above which actually helps a lot, thanks!

**Reply by wooorm** [reply]
> As to my use case, it's pretty simple. I have a markdown editor on the client. I just want to load markdown files and preserve the structure, so ...

If you want the raw text, then why are you parsing it with a markdown parser? If you want the plain text, get the file and *don’t* pass it to a markdown parser, but display that raw file instead?

**Reply by nicknikolov** [reply]
No, I do parse and render the AST rather than the raw text. But Christian is right I think it's up to the transformer/client. For example, since I render in HTML so I would either have to add bottom margin to show the split or something like that (maybe add an extra p tag).  

I think the confusion came from the fact that I had one conceptual understanding of what a paragraph _is_ based on HTML and that is slightly different with Markdown.

In HTML you can do
```
<p>foo</p>
<p></p>
<p>bar</p>
<p></p>
```
and this is what I wanted my pipeline to produce. But it turns out those empty lines signify the split of the paragraph in Markdown so in the end you get only two. So I will need to think about how to manage that difference.

**Reply by wooorm** [reply]
Nice. Sounds like you’ve got a good understanding of what’s going on and how to solve it.

One thing though: In HTML you also shouldn’t add empty paragraphs. HTML has *meaning*. So this might be more of a CSS thing, which adds *presentation*!
